schema/image.go

```diff
@@ -1,31 +1,84 @@
 package schema
 
 import (
+	"fmt"
 	"strings"
 )
 
+// BuildFormat defines the docker image tag format that is used during the build process
 type BuildFormat int
 
 // DefaultFormat as defined in the YAML file or appending :latest
 const DefaultFormat BuildFormat = 0
 
+// SHAFormat uses "latest-<sha>" as the docker tag
 const SHAFormat BuildFormat = 1
 
+// BranchAndSHAFormat uses "latest-<branch>-<sha>" as the docker tag
 const BranchAndSHAFormat BuildFormat = 2
 
+// DescribeFormat uses the git-describe output as the docker tag
+const DescribeFormat BuildFormat = 3
+
+// Type implements pflag.Value
+func (i *BuildFormat) Type() string {
+	return "string"
+}
+
+// String implements Stringer
+func (i *BuildFormat) String() string {
+	if i == nil {
+		return "latest"
+	}
+
+	switch *i {
+	case DefaultFormat:
+		return "latest"
+	case SHAFormat:
+		return "sha"
+	case BranchAndSHAFormat:
+		return "branch"
+	case DescribeFormat:
+		return "describe"
+	default:
+		return "latest"
+	}
+}
+
+// Set implements pflag.Value
+func (l *BuildFormat) Set(value string) error {
+	switch strings.ToLower(value) {
+	case "", "default", "latest":
+		*l = DefaultFormat
+	case "sha":
+		*l = SHAFormat
+	case "branch":
+		*l = BranchAndSHAFormat
+	case "describe":
+		*l = DescribeFormat
+	default:
+		return fmt.Errorf("unknown image tag format: '%s'", value)
+	}
+	return nil
+}
+
 // BuildImageName builds a Docker image tag for build, push or deploy
-func BuildImageName(format BuildFormat, image string, SHA string, branch string) string {
+func BuildImageName(format BuildFormat, image string, version string, branch string) string {
```

Comment:
@alexellis i am curious how you would feel about adding another option that allows us strip the `latest`.  In my ideal workflows I want images that are tagged `foo:v1.0.1` or `foo:a7afdl` not `foo:latest-v1.0.1`

Reply:
The text `latest` is the value from the stack.yml file:

```yaml
functions:
  git-tar:
    lang: go
    handler: ./git-tar
    image: git-tar:0.11.0
  buildshiprun:
    lang: go
    handler: ./buildshiprun
    image: buildshiprun:0.5.0

# repo: 0.9.0
```


It is effectively the static image tag as specified in the file, which may well be different to the tag of the Git repository. Take OpenFaaS cloud for instance. Each function has an independent version of the Git repo release.